reactor-netty-core/src/main/java/reactor/netty/resources/DefaultLoopResources.java

```diff
@@ -201,17 +203,28 @@ EventLoopGroup cacheNioSelectLoops() {
 
 	@SuppressWarnings("FutureReturnValueIgnored")
 	EventLoopGroup cacheNioServerLoops() {
-		EventLoopGroup eventLoopGroup = serverLoops.get();
-		if (null == eventLoopGroup) {
+		final EventLoopGroup eventLoopGroup = serverLoops.get();
+
+		if (null != eventLoopGroup) {
+			return eventLoopGroup;
+		}
+
+		if (eventLoopGroupCreationInProgress.compareAndSet(false, true)) {
+			if (null == serverLoops.get()) {
 			EventLoopGroup newEventLoopGroup = new NioEventLoopGroup(workerCount,
 					threadFactory(this, "nio"));
-			if (!serverLoops.compareAndSet(null, newEventLoopGroup)) {
+				if (serverLoops.compareAndSet(null, newEventLoopGroup)) {
+					eventLoopGroupCreationInProgress.set(false);
+				} else {
 				//"FutureReturnValueIgnored" this is deliberate
 				newEventLoopGroup.shutdownGracefully();
 			}
-			eventLoopGroup = cacheNioServerLoops();
 		}
-		return eventLoopGroup;
+		} else {
+			while (null == serverLoops.get()) {}
```

Comment:
rather than using recursion as the last step, you could reuse `eventLoopGroup` (making it not final) and continuously try to reload it here, then return `eventLoopGroup` at the end? (`newEventLoopGroup` can also be replaced by `eventLoopGroup` in that case)

I also wonder if anything can go wrong in the `NioEventloopGroup` ctor, which would mean the other thread would get stuck in this infinite loop (ie can this be proven to be guaranteed to make progress?)

Reply:
I try to avoid mutability as much as possible, that's the reason.  But, I am not objecting your suggestion.  Also, yes, there is risk.  Depending on how the `warmup` api discussion goes with @violetagg, we may close this PR.  If not, I will make the suggestions you made and also will add ways to prevent infinite loop.